[Poker Game App step6] 포커게임 승자 찾기

dev-Lena opened the pull request:
## 현재 상황

![](https://i.imgur.com/6yeHofa.jpg)

1. ViewController에서 StackView와 SegmentedControl를 분리해봤지만 뷰 계층에는 존재하는데 SegmentedControl가 simulator에 보이지 않습니다. 

2. 승자를 찾는 과정
2-1.참가자(플레이어들 + 딜러)가 갖고있는 카드들 중에 어떤 조합(combination)이 있는지 확인하고 조합에 따라 점수를 매깁니다. 
2-2.가장 높은 점수를 갖은 참가자가 두 명 이상이 경우(동점자), 조합을 이룬 카드들 중 더 높
2-3. 은 Rank를 갖고있는 참가자에게 추가 점수가 부여됩니다. 
2-4.이후 가장 높은 점수의 참가자가 winner로 선정됩니다. 

## 추가로 구현해야하는 부분
1. 페어가 없는 경우에 대한 부분을 추가해야합니다.
2. 화면에 승자 표시를 위한 이미지 뷰를 추가하고, 승자를 표시하는 부분을 추가로 구현해야합니다.

## 고민한 점 및 궁금한 점

1. (현재 상황 1번) 이전 step에서 view의 생성과 세팅부터 ViewController와 분리해보라는 피드백을 주셔서 시도해봤습니다. 그런데 화면에 왜 나타나지 않는지 모르겠습니다... 

2. Array, Dictionary, Set 중 어떤 Collection에 값을 담을지 계속 고민해봤습니다. 사실 같은 숫자를 찾는다고 해서 Set의 intersect을 사용해서 해보려고 했었는데, 이리저리 해보다가 결국 이번 스탭에서 Set이 적절하지 않은 것같아서 Dictionary와 Array를 사용했습니다. 지금도 크게 효율적이라고 생각이 되지 않는데, 혹시 더 나은 방법이 있다면 추후 수정하겠습니다!

3. 어제 부터 이 에러를 많이 만났는데 이 에러가 어떤 에러인지 잘 모르겠습니다.
Build도 success하고 simulator도 Run이 되지만 test 코드를 실행시키거나 test 파일까지 build하면 build fail이 뜹니다.

다른 swift 파일에 클래스와 이넘이 있는데 어제부터 이런 에러가 생겼다가 없어졌다가 하고 있습니다. 왜 그런지 이유를 잘 모르겠습니다...
![](https://i.imgur.com/17aSHvP.png)
![](https://i.imgur.com/OFD47T2.png)

혹시 몰라서 이런 노력을 해봤습니다.
- 각 swift 파일을 테스트 타겟 설정해줬습니다.
- Team을 none으로 하지 않고 제 애플 계정 personal로 바꿔줬습니다.

dev-Lena committed:
ViewController에 Shake 이벤트 메소드 추가

앱에서 Shake 이벤트를 발생하면 랜덤 카드를 다시 섞고 다시 동작하도록 하는 메소드를 추가했습니다.

dev-Lena committed:
Card 클래스 extension 추가

Card의 Rank를 알파벳이 아니라 숫자로 반환하는 numberDescription을 추가했습니다.

dev-Lena committed:
GameStackView 클래스 추가

ViewController에서 참가자(플레이어와 딜러)의 라벨과 카드들을 담는 StackView에 대한 클래스를 생성했습니다.

dev-Lena committed:
GameStackView에 setConstraintOfView(related:, related:) 추가

dev-Lena committed:
CardsStackView 클래스 생성

dev-Lena committed:
ParticipantStackView 클래스 생성

dev-Lena committed:
SegmentedControlProtocol 생성

dev-Lena committed:
gameModeSegmentControl 클래스 생성

dev-Lena committed:
NumbersOfPlayersSegmentControl 클래스 생성

dev-Lena committed:
ViewController view 프로퍼티 수정

StackView와 SegmentedControl에 대한 프로퍼티 부분을 수정했습니다.

dev-Lena committed:
ViewController에서 viewDidLoad()메소드 수정

dev-Lena committed:
ViewController SegmetnedControlProtocol 채택한 extension 추가

dev-Lena committed:
ViewController에 setPlayersCards(in:) 수정

dev-Lena committed:
ViewController에 setDealersCards(in:) 수정

dev-Lena committed:
ViewController에 addDealerCards(of:) 수정

dev-Lena committed:
ViewController에서 setConstraintOfView(of:) 삭제

dev-Lena committed:
Combination 이넘 추가

카드 조합에 대한 종류 타입을 나타내는 Combination 이넘을 추가했습니다.
각 타입의 rawValue는 조합에 따른 점수입니다.
타입에 맞는 점수를 알려주는 tellScore()를 추가했습니다.

dev-Lena committed:
HandCombination 클래스 생성

dev-Lena committed:
HandCombination 클래스에 checkStraight(of:) 추가

카드 조합 중 카드 rank가 연속된 5개의 수인 카드가 있는지(스트레이트) 확인하는 메소드를 추가했습니다.

dev-Lena committed:
CheckStraight()에 대한 테스트 케이스 추가

dev-Lena committed:
HandCombination 클래스에 checkPair(of:) 추가

dev-Lena committed:
CheckPair()에 대한 테스트 케이스 추가

dev-Lena committed:
Utility 클래스 생성 및 convertToNumber(text:) 추가

dev-Lena committed:
테스트 케이스 추가 : testConvertToCombination()

dev-Lena committed:
ScoreBoard 클래스 생성

ScoreBoard 클래스가 초기화 될 때, 참가자 라벨을 key로 갖고 각각 value로 0점을 갖고있도록 했습니다.
doInitialSetup(), updateScores(of:), tellHighestScore(), addScore(to:) 메소드를 추가했습니다.

dev-Lena committed:
ScoreBoard 클래스에 동점자를 처리하는 메소드 추가

dev-Lena committed:
테스트 케이스 추가 : ScoreBoard에서 점수 추가와 동점자 처리 메소드 테스트

dev-Lena committed:
테스트 케이스 추가 : testScoreBoardInitialSetUp()

dev-Lena committed:
WinnerDiscriminator 클래스 생성

참가자들이 갖고있는 카드에서 조합을 찾고 점수를 매겨 승자를 찾아내는 클래스를 생성했습니다.

dev-Lena committed:
WinnerDiscriminator 클래스에 카드의 조합과 승자를 찾는 메소드 추가

dev-Lena committed:
테스트 케이스 추가 : WinnerDiscriminator 클래스의 메소드를 테스트하는 케이스 추가

godrm commented on CardGameApp/CardGameApp/View/SegmentedControl.swift:
```diff
@@ -0,0 +1,81 @@
+    
+    private var gameMode = GameMode.fiveCardStud
+    let gameModes: [String] = ["5 Cards","7 Cards"]
+    var gameModeSegmentControl = UISegmentedControl()
```
UISegmentedControl 을 상속 받았는데 내부에서 또 UISegmentedControl 인스턴스를 만들고 있네요. 뷰 클래스는 그 자체가 뷰에 그리는 역할을 하는거고, 내부에 다른 속성이 있으면 하위 뷰를 의미합니다. 후자를 표현하려는 게 맞나요?

godrm commented on CardGameApp/CardGameApp/View/SegmentedControl.swift:
```diff
@@ -0,0 +1,81 @@
+    }
+    
+    convenience init() {
+        self.init(frame: CGRect(x: 0, y: 0, width: 0, height: 0))
```
뷰 객체를 만들때 frame 정보는 매우 중요합니다. 이렇게 만들고 싶으면 `init(frame: CGRect.zero)` 를 활용하는 게 어떨까요

godrm commented on CardGameApp/CardGameApp/View/SegmentedControl.swift:
```diff
@@ -0,0 +1,81 @@
+        gameModeSegmentControl = UISegmentedControl(items: gameModes)
+        self.addTarget(self, action:
+            #selector(valueDidChange),
+                       for: .valueChanged)
```
target / action을 자기 자신으로 지정했네요. 타겟을 지정하는 것은 외부에서 이벤트를 받을 객체가 지정하는 게 좋습니다. 뷰는 누가 자기 객체가 될지 모르니까요. Delegate 패턴으로 만들고 싶었다면 Delegate 객체에게 직접 보내면 됩니다. Target / Action은 Delegate 없이 처리할 수 있는 방식입니다.

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -9,90 +9,37 @@
+        self.view.addSubview(gameModeSegmentedControl)
+        gameModeSegmentedControl.setSegmetedControl(in: self.view, offset: 15)
+        
+        self.view.addSubview(numbersOfPlayersSegmentedControl)
```
뷰 인스턴스는 frame 속성이 매우 중요합니다. frame 정보를 넣어주지 않았으니 뷰가 표시되지 않는 게 당여한 것 같습니다. 
frame 정보를 원하는 위치로 지정해보세요.

godrm commented on CardGameApp/CardGameApp/View/SegmentedControl.swift:
```diff
@@ -0,0 +1,81 @@
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        super.init(coder: aDecoder)
```
다른 뷰도 마찬가지지만, 뷰를 서브 클래스로 만들 때 init 함수는 중요한 역할을 합니다.
init(frame:) 과 init(coder:) 둘 다 어떤 것을 호출해도 동일하게 초기화를 해야 합니다.
그래야 동일한 동작을 구현할 수 있다고 신뢰할 수 있습니다. 두 메소드가 어떤 경우에 각각 불리는지 학습해보세요.

godrm commented on CardGameApp/CardGameApp/Model/WinnerDiscriminator.swift:
```diff
@@ -0,0 +1,91 @@
+    var participantsCardsDictionary : [String: [Int]] = [:]
+    var participantsCombinedCardsDictionary : [String: [Int]] = [:]
+    var allParticipants = [String]()
+    var scoreBoard = ScoreBoard(with: [])
```
WinnerDiscriminator 라는 별도 클래스를 만들었네요. 근데 Game보다도 더 상위 개념으로 설계한 것은 아쉽습니다. 
이 클래스는 Game 속성이나 Players, Dealer 모두와 내부 카드까지 알아야 하고
그럴려면 내부 속성을 가져오거나 탐색, 비교하는 로직을 모두 포함하게 됩니다. 
다른 객체들과 강하게 결합되어야만 합니다. 객체지향 설계에서는 다른 객체와 강하게 결합되는 구조를 지양하는 게 좋습니다. 그래야 변경할 때 유연해지니까요. 
여기에 있는 상당수는 Game 이나 Participants 에서 처리하는 게 좋습니다.

godrm commented on CardGameApp/CardGameApp/Model/Utility.swift:
```diff
@@ -0,0 +1,15 @@
+class Utility {
+    func convertToNumber(text : String) -> Int{
+        return Int(text)!
+    }
```
이렇게 유틸리티 함수를 모아 놓는 경우는 매개변수로 넘어가는 타입에 주목해보세요.
그냥 String 타입을 확장하는 게 낫지 않을까요?

godrm commented on CardGameApp/CardGameApp/Model/ScoreBoard.swift:
```diff
@@ -0,0 +1,72 @@
+     func compareHighestRank(between tiebreakers : [String:Int], ranksOfCombinedCards : [String:[Int]]){
+         var comparingCardRank = 0
+         for name in tiebreakers.keys{
+             if let combindedCards = ranksOfCombinedCards[name]?.sorted(by: >){
```
이런 경우는 if let 보다 guard let 을 써서 early exit으로 구현하면 들여쓰기를 줄일 수 있습니다.

godrm commented on CardGameApp/CardGameApp/Model/ScoreBoard.swift:
```diff
@@ -0,0 +1,72 @@
+class ScoreBoard{
+    
+    let participants : [String]
+    var scoreBoard = [String: Int]()
```
항상 private 을 우선적으로 고려하세요.

godrm commented on CardGameApp/CardGameApp/Model/ScoreBoard.swift:
```diff
@@ -0,0 +1,72 @@
+    let participants : [String]
+    var scoreBoard = [String: Int]()
+    init(with participants: [String]) {
+        self.participants = participants
```
이렇게 참자가를 다시 목록으로 받아야 한다면, 여기 기능들은 Participants 내부에서 구현하는 게 어땠을까요?

godrm replied:
스코어 보드가 participants 보다 상위 개념인가요? 아니면 대등한 관계인가요? 
각각의 스코어를 갖고 있어야 한다면 participant 각자 스코어를 갖고 있으면 participants 와 거의 비슷한 관계가 되지 않을까요?

godrm commented on CardGameApp/CardGameApp/Model/HandCombination.swift:
```diff
@@ -0,0 +1,111 @@
+                    duplicationRemovedCards.remove(at: cardIndex)
+                }
+            }
+        }
```
간단히 배열 내에서 같은걸 제거하는 거라면 내부에서 Set을 쓰는게 어땠을까요

godrm commented on CardGameApp/CardGameApp/Model/HandCombination.swift:
```diff
@@ -0,0 +1,111 @@
+    var combinedCardsRank = [Int]() // 조합을 이룬 카드들의 Rank
+    var combinationTypes = [Int]() // 참가자가 갖고있는 카드들 중 각 조합을 이루기 위한 카드는 몇개였는지 (ex - Triple = 3, Straight = 5)
+    var isOnePair : Bool = false
+    var typesOfCombination = [Combination]() // 참가자가 갖고있는 카드들이 어떤 조합이 나왔는지
```
항상 Private을 고려하세요. 
HandCombination 은 어느 객체과 결합되어야 하는 걸까요? 
이미 만들어진 클래스가 있을 때 새로운 클래스를 만드는 경우는 어떤 조합으로 결합할껀지, 어느 클래스 하위에 포함할건지 고민해봐야 합니다. 
저는 카드 목록을 갖고 있는 대상과 결합하면 좋겠다고 생각합니다.

godrm reviewed (commented):
현재 상황에서 중요한 포인트가 몇 개 있네요. 그걸 해결해야 마무리가 될 것 같아서 피드백을 남깁니다.

dev-Lena committed:
SegmentedControl 클래스 수정

각각 클래스로 만들었던 두 SegmentedControl View를 하나의 클래스로 합쳤습니다.
SegmentedControl의 Constraint를 설정하는 SetUpView()를 수정했습니다.

dev-Lena committed:
바뀐 SegmentedControl 클래스에 맞게 ViewController 수정

ViewController에 프로퍼티로 선언되어있던 부분을 수정하고 불필요한 부분은 삭제했습니다.
SegmentedControlProtocol을 채택한 extension 부분에서 선택사항이 바뀔 때마다 반영하여 새로운 포커게임을 시작하도록 했습니다.

dev-Lena committed:
배열에서 중복제거하는 메소드 수정

반복문으로 현재 인덱스에 있는 value와 다음 인덱스에 있는 index를 비교해서 중복값을 제거하는 방식에서 Set을 통해 중복값을 제거하는 방법으로 수정했습니다.

dev-Lena committed:
Players에서 PokerPlayer로 클래스 이름 변경

dev-Lena committed:
PokerPlayer 클래스의 프로퍼티 name에 관련된 메소드 추가

dev-Lena committed:
Players에서 Participants로 클래스 변경

기존에 Player 배열을 프로퍼티로 갖고있고 이와 관련된 메소드가 있던 Players 클래스를 Player 배열과 Dealer를 갖고있고 이와 관련된 메소드가 있는 Participants 클래스로 바꿨습니다.

dev-Lena committed:
Card extension에 Card의 Rank를 Int로 반환하는 클로저 변수 추가

dev-Lena committed:
PokerGame 클래스에 프로퍼티였던 Players를 Participants로 변경

dev-Lena committed:
ViewController에서 Dealer 관련 메소드 삭제

Players와 Dealer를 분리해서 작업하지 않고 Participants로 작업하도록 변경

dev-Lena committed:
PokerPlayer에 HandCombination 프로퍼티 추가

카드 배열을 갖고 있는 PokerPlayer가 수중의 카드 중 조합을 찾는 HandCombination을 갖고있도록 추가했습니다.

dev-Lena committed:
PokerPlayer클래스 수정

각 참가자(플레이어, 딜러)가 조합을 체크하는 HandCombination을 가지고 있습니다.
자기 자신이 수중에 들고있는 조합을 체크하고 조합을 이룬 카드의 랭크들을 배열에 저장합니다.
조합 타입도 배열로 저장해서 프로퍼티로 가지고 있도록 수정했습니다.

dev-Lena committed:
HandCombination 클래스 수정

배열이 nil일 경우를 처리하는 dealWithNilValue()를 추가했습니다.
submitCheckResult(of:)에서 조합을 체크하고 조합을 이룬 카드의 Rank 배열과 조합타입 배열을 담은 배열을 튜플로 반환하도록 수정했습니다.

dev-Lena committed:
PokerGame 클래스 수정

WinnerDiscriminator클래스에서 승자를 찾던 부분을 PokerGame으로 옮겼습니다.

dev-Lena commented:
### 수정한 점
1. ViewController에서 SegmentedControl과 StackView는 클래스를 따로 만들었습니다. 
2. Segmented Control 뷰가 화면에 보이지 않던 점을 해결했습니다.
3. ScoreBoard 클래스를 삭제하고 PokerGame에 딕셔너리 프로퍼티로 추가했습니다.
4. 이전에는 승자를 찾기위해 WinnerDiscriminator 라는 별도 클래스를 만들었었는데, 이 클래스를 삭제하고 안에 있던 프로퍼티나 메소드들을 PokerGame이나 Participants 클래스에 나눴습니다.

내일 아직 승자에게 메달 이미지를 달 예정입니다.

### 궁금한 점
위에서 나왔던 에러들은 대부분 해결했는데 이 에러는 어떤 문제인지 모르겠습니다...
![](https://i.imgur.com/FVP9IlI.png)

godrm commented on CardGameApp/CardGameApp/Model/Card.swift:
```diff
@@ -63,4 +63,7 @@ extension Card {
         return "\(rank.description)"
     }
+    var cardNumber: Int{
+        return rank.rawValue
```
타입을 새로 선언했는데 이렇게 내부 속성을 그대로 가져가는 방법을 지양하세요. 
값을 가져가서 처리하는 동작을 생각해보세요. 
비교하는 거라면 여기에 그 비교하는 동작을 가져와서 구현하세요.

godrm commented on CardGameApp/CardGameApp/Model/HandCombination.swift:
```diff
@@ -105,7 +101,23 @@ class HandCombination{
+        typesOfCombination += pairCombination
+        dealWithNilValue()
+        Array(Set(typesOfCombination))
+        Array(Set(combinedCardsRank))
```
이 코드는 뭘까요? 다른 객체를 만들고 전혀 사용하지 않는 것 같은데요?

dev-Lena replied:
아 그러네요! 수정하겠습니다!

godrm commented on CardGameApp/CardGameApp/Model/PokerGame.swift:
```diff
@@ -54,26 +59,68 @@ class PokerGame {
+        breakTie()
+        let winner = scoreBoard.max(by:){ participant, anotherParticipant in participant.value < anotherParticipant.value}?.key
+        guard let theWinner = winner else { return ""}
+        return theWinner
```
winner 리턴타입이 왜 String 이어야 하나요? participant 객체를 참조해도 되지 않을까요?

dev-Lena replied:
ScoreBoard의 key에 들어가는 타입을 String이 아니라 participant 객체로 변경하겠습니다!

godrm commented on CardGameApp/CardGameApp/Model/PokerGame.swift:
```diff
@@ -54,26 +59,68 @@ class PokerGame {
+                let score = combination.tellScore()
+                addScore(to: name, extraScore: score)
+            }
         }
```
Participants 가 너무 수동적인 객체가 되어 버렸네요. 
여기 있는 상당수들은 Participants가 책임지고 구현해도 될 것 같아요. 
어차피 Game에 있는 속성이나 값을 업데이트 해야 하는게 아니라면 
allParticipants.updateScore() 라고 호출해도 되는게 아닐까요?
PokerGame > Participants > Participant 포함관계에 따라서 상위에서 하위로 시켜야 하는 동작을 구분할 필요가 있습니다.
모든 것을 클로저로 구현해야 하는 건 아닙니다. 매개변수와 리턴값을 보면 어떤 객체와 더 관련있고 응집되어 있어야 하는지 힌트가 됩니다.

dev-Lena replied:
넵!! 힌트 감사합니다! 이 부분 다시 고민해서 수정하겠습니다!

godrm commented on CardGameApp/CardGameApp/Model/PokerPlayer.swift:
```diff
@@ -0,0 +1,71 @@
+    }
+}
+
+class Player : PokerPlayer {
```
Player 세부 기능이 없다면 선언할 필요가 없습니다.
 그냥 PokerPlayer를 쓰면 되는거죠.

godrm commented on CardGameApp/CardGameApp/Model/WinnerDiscriminator.swift:
```diff
@@ -32,34 +30,26 @@ class WinnerDiscriminator{
             participantsList.append("Player\(playerNumber)")
             playerNumber += 1
         }
         participantsList.append("Dealer")
```
이 정도 깊이로 접근하는 것은 좋지 않은 징조입니다. 
PokerGame 에도 participants 가 있는데 여기도 allParticipants 가 있네요.
PokerGame 에 participants 내부에 있는 player까지 접근할 수 있는 클로저가 있어서 그걸로 동일한 allParticipants까지 만들 필요는 없죠. 그대로 participants를 쓰면 되는거니까요.

dev-Lena replied:
아 이 클래스는 사용하지 않는데 삭제하는걸 잊었네요! 이 클래스에 있던 내용은 PokerGame클래스와 Participants 클래스로 옮겼습니다.

godrm commented on CardGameApp/CardGameApp/Model/WinnerDiscriminator.swift:
```diff
@@ -32,34 +30,26 @@ class WinnerDiscriminator{
             }
             participantsCardsDictionary[allParticipants[participantIndex]] = onePlayerCards
             participantIndex += 1
         }
```
이렇게 상위 객체에서 하위 객체 속성까지 접근해서 구현하기보다는 하위 객체들에게 각자 반복적인 일을 하도록 구조를 만드는 게 좋습니다. 
지금은 카드게임한테 각자 플레이어를 알려주고, 플레이어에게 카드 보여주고, 카드 번호까지 알아내서 조합하고 있는 것을 4단계 위 상위 모듈에서 한꺼번에 처리하고 있습니다. 

이것보다는 상위 모듈은 바로 아래 카드 게임한테 결과를 알려줘. 라고 하면 
카드 게임이 모든 players들에게 각자 카드를 알려줘. 하고
그걸 받아서 취합해서 상위로 return 해주는 방식으로 단계적으로 일을 하도록 구현하세요.

클로저를 활용하라고 했던 것은 이미 있는 객체를 반복적으로 탐색하는 경우에만 활용하는 게 좋습니다. 
그 외에는 상위 모듈이 로직을 갖고 있을 필요가 없습니다!!

godrm commented:
> ### 궁금한 점
> 위에서 나왔던 에러들은 대부분 해결했는데 이 에러는 어떤 문제인지 모르겠습니다...
> ![](https://camo.githubusercontent.com/859437b7ecf2cec13292c32e6e0b9bd63cbb8044/68747470733a2f2f692e696d6775722e636f6d2f46565039496c492e706e67)

이 문제는 테스트 타깃에 설정이나 특정한 파일이 문제가 있는 것 같습니다. 점심 먹고 한 번 같이 봐요.

dev-Lena committed:
WinnerDiscriminator 클래스 삭제

dev-Lena committed:
Player 클래스 삭제

dev-Lena committed:
PokerGame에 scoreBoard 딕셔너리의 키 타입을 PokerPlayer로 바꾸고 Hashable 구현

dev-Lena committed:
승자에게 메달이미지 추가

dev-Lena commented:
### 수정한 점과 현재 상황
1. ScoreBoard key의 타입을 Hashable을 이용해서 String이 아닌 PokerPlayer 타입으로 수정했습니다 
2. 승자의 인덱스 번호를 가져오고 승자 메달 이미지를 출력합니다. 그런데 승자 옆에 달리지는 않습니다.. :(

### 궁금한 점
1. Card의 Rank.rawValue를 그대로 가지고 오지 않고 extension한 부분에서 직접 비교하는 동작을 가지고 와서 구현하라는 피드백을 주셨는데,
제가 한 방법은 card들의 Rank를 rawValue(숫자)로 받아와 배열에 넣은 다음 조합을 찾는 방식인데 Card 클래스에서 어떻게 비교하면 좋을지 고민해봤는데 감을 잘 못잡겠습니다.. :(

2. self.view 안에 gameStack이 있고, 그 안에 각 참가자의 라벨과 카드들이 들어있는 oneParticipantInfoStack이 있습니다.
winner인 참가자가 gameStack 몇 번째에 있는지 winnerPosition(index)를 받아왔습니다. (findWinner())
승자 메달 이미지뷰의 topAnchor와 bottomAnchor를 그 위치의 oneParticipantInfoStack의 topAnchor와 bottomAnchor와 같게 했습니다.
그런데 아래와 같이 위치가 적용되지 않았는데 어떤 부분 때문인지 잘 모르겠습니다... :((

### 현재 상황 이미지 
![](https://i.imgur.com/JnMhbW3.png)

godrm commented:
> ### 궁금한 점
> 1. Card의 Rank.rawValue를 그대로 가지고 오지 않고 extension한 부분에서 직접 비교하는 동작을 가지고 와서 구현하라는 피드백을 주셨는데,
>    제가 한 방법은 card들의 Rank를 rawValue(숫자)로 받아와 배열에 넣은 다음 조합을 찾는 방식인데 Card 클래스에서 어떻게 비교하면 좋을지 고민해봤는데 감을 잘 못잡겠습니다.. :(

핵심은 Card.Rank.rawValue로 조합을 찾는 것을 카드로 조합을 찾으면 된다는거죠. 
조합을 찾을 때 어떻게 비교하고 있나요? 
그 비교하는 방법을 Card 와 Rank에서 메소드로 제공하면 판단할 수 있지 않을까요?
상위 모듈에서 값을 가져와서 다른 데이터 구조를 만들어서 비교하지 말고,
하위 모듈(Card 또는 Rank) 그 자체에서 비교하고 판단하도록 해야 객체 역할이 나눠집니다. 

> 2. self.view 안에 gameStack이 있고, 그 안에 각 참가자의 라벨과 카드들이 들어있는 oneParticipantInfoStack이 있습니다.
>    승자 메달 이미지뷰의 topAnchor와 bottomAnchor를 그 위치의 oneParticipantInfoStack의 topAnchor와 bottomAnchor와 같게 했습니다.

승자 메달 이미지뷰를 어느 뷰 아래에 addSubview를 하나요? 
뷰의 자표 시스템은 추가하는 superview 아래 상대좌표가 됩니다. 

self.view 아래 gameStack 아래 oneParticipantInfoStack 이렇게 되면 어디와 동일해야 할까요?

godrm commented on CardGameApp/CardGameApp/View/StackView.swift:
```diff
@@ -64,12 +64,30 @@ class ParticipantStackView : UIStackView {
+    }
+    
+    required init(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
ParticipantSectionStack처럼 뷰 객체를 서브클래스로 만들 때는 init 메소드를 어떤 것을 쓰더라도 동일한 결과가 나오도록 만드는 게 좋습니다. init(coder:)를 당장 사용하지 않더라도 이렇게 내버려두지 마세요. 다른 init과 동일하게 초기화가 되도록 만들어야 합니다.

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -28,16 +30,18 @@ class ViewController: UIViewController {
+        pokerGame.shuffleWholeCardDeck()
+        self.view.addSubview(gameStackView)
+        gameStackView.setConstraintOfView(related: self.view, related: segmentedControl)
+        setParticipantsCards(in : gameStackView)
```
뷰를 배치하는 게 꽤나 소모적인 일이라서 remove 하고 add 하고 반복하는 게 좋은 경우도 있고,
이미 레이아웃을 다 잡아놓고 감춰놓고 있다가 보여주기만 하는 경우도 있습니다. 
메달을 보여주는 방식을 조금 쉽게 다 배치해놓고 위치에 따라 보여주거나 감추는 건 어떨까요?

dev-Lena committed:
StackView 클래스 required init(coder:) 수정

dev-Lena committed:
Rank 이넘과 Card 클래스 Comparable 프로토콜 채택 및 비교연산자 재정의

dev-Lena committed:
HandCombination 클래스 수정

Card의 Rank를 비교하는 부분에서 Int 타입을 사용했던 부분을 Card 타입으로 수정했습니다.
checkStraight(of:) 메소드를 Card 클래스에서 재정의한 '<'를 사용해 수정했습니다.

dev-Lena committed:
Participants 클래스에서 giveExtraScoreForHigherRank(addScore:) 수정

dev-Lena committed:
불필요한 코드 삭제

dev-Lena committed:
우승자 옆에 메달 이미지 추가

모든 우승자에게 메달 이미지를 다 배치해놓고 우승자의 위치(subview의 index)에 있는 메달만 보여지도록 수정했습니다

dev-Lena committed:
README.md에 구현 사진 업데이트

dev-Lena commented:
### 수정한 점
- [x]  View 객체를 서브 클래스로 만들 때 init(code:)를 다른 init과 동일하게 초기화 되도록 수정했습니다!
- [x]  Card의 Rank.rawValue를 그래도 가지고 오지 않도록 수정했습니다. Card 클래스 extension한 후 Compable 프로토콜을 채택해 ==와 < 재정의했습니다!
- [x]  모든 우승자에게 메달 이미지를 다 배치해놓고 우승자의 위치(subview의 index)에 있는 메달만 보여지도록 했습니다!

### 결과화면
![image](https://user-images.githubusercontent.com/52783516/76687164-bc4ac180-6664-11ea-8ea7-42fdbfbd25a8.png)

godrm commented:
수고하셨습니다!
